File: cPP_java_py_js-DSA/practice.py
#Function to find the nth fibonacci number 
def fibbonacci(n):
    if (n<=1): return n
    else : return fibbonacci(n-1)+fibbonacci(n-2)
#Function to print the fibonacci series 
def fibPrint(n):
    for i in range(5):
        print(fibbonacci(i))
#Function for finding factorial of number
def factorial(n):
    if (n<=1): return 1 
    else :return n*factorial(n-1)

#Function to reverse a string 
def reverseString(s):
    #string are immutable in pyton
     # Convert string to list of characters
    char_list = list(s)
    
    n = len(char_list)
    i = 0 
    j = n - 1
    
    while i < j:
        # Swap characters
        temp = char_list[i]
        char_list[i] = char_list[j]
        char_list[j] = temp
        i += 1
        j -= 1
    
    # Join list back into a string
    return ''.join(char_list)
   
#Functino for finding mimimum and maximum element in the array 
def minMax(arr ):
    n = len(arr)
    mini= arr[0]
    maxi = arr[1]
    for i in range (n):
        if arr[i]<mini:
            mini = arr[i]
        if arr[i]>maxi:
            maxi = arr[i]
    ans = [mini,maxi]
    return ans

# Non returning function and with no paramters
def greet():
    print("Good Morning Mitron")

# returning function with passing parameters 
def add(a,b ):
    return (a+b)


print('hello')
#comments in python
print("\nlearning the basics of python")

#important thing is indendation (space from starting )

#data types , variable decalration and initialization 
a=10 # we cannot print the value of variable without intialize it  
print(a)

#in python we explictly do not have any data types like int , float , string , bool (we can give any kind of value to a variable similiarily to javascript)
a ="ramesh pal" # we can change the variable value to one data type to other 
print(a)

# User Input 
a = input(" : Enter value of a ")
print(a)

# 3 time string literal make multi line comment 
"""
FUNCTION DECLARATION  , CALLING , PARAMTER PASSING

"""
greet() # functiona calling 
print(add(10,20))

"""
IF -ELSE STATEMENTS      
"""
a = int(input("enter your age ")) # need to explictly convert from string to int 
if (a>18): 
    print("your are a Adult")
elif( a>=0 and a<=5):
    print("child")
else : 
    print("you are a Minor")

"""
LOOPS CONCEPT IN PYTHON
"""
# for loop

for i in range (1,11):
    print(2*i," ")

# while loops 
i=2

while i<=10:
    print(i)
    i+=2

#do while concept doesn't exists in python

"""
ARRAY CONCEPT IN PYTHON 
"""
#array /List declaration 
a =[2,4,1,10,100]
#size finding
n = len(a)
print(n)
print(a)

#array passing as argument
ans = minMax(a)
print(ans)

#string passing as arguement 
s = "yash Munde"
print(len(s))
print(s)
rev =reverseString(s)
print(rev)

#recursion implementation

fact  = factorial(5)
print(fact)

fibPrint(5)

File: cPP_java_py_js-DSA/practice.cpp
#include <bits/stdc++.h>
using namespace std;
// function to find the factorial of a number
int factorial(int n)
{
    // base case
    if (n <= 1)
    {
        return 1;
    }
    // recursive case
    return n * factorial(n - 1);
}

// function to print fibonacci series
int fibonacci(int n)
{
    if (n <= 1)
        return n;
    return fibonacci(n - 1) + fibonacci(n - 2);
}
void fiboPrint(int n)
{
    for (int i = 0; i < n; i++)
    {
        cout << fibonacci(i) << " ";
    }
}
// function for reverse a string
string reverseString(string str)
{
    int n = str.size();
    int i = 0, j = n - 1;
    while (i < j)
    {
        swap(str[i], str[j]);
        i++;
        j--;
    }
    return str;
}
// function for vowel , consonent count
void countVowelConsonent(string str)
{
    int vowel = 0, consonent = 0, whiteSpace = 0;
    for (int i = 0; i < str.size(); i++)
    {
        if (str[i] == 'a' || str[i] == 'e' || str[i] == 'i' || str[i] == 'o' || str[i] == 'u')
        {
            vowel++;
        }
        else if (str[i] == 'A' || str[i] == 'E' || str[i] == 'I' || str[i] == 'O' || str[i] == 'U')
        {
            vowel++;
        }
        else if (str[i] == ' ')
            continue;
        else
        {
            consonent++;
        }
    }
    cout << "length of string is " << str.size() << endl;
    cout << "Vowel : " << vowel << " Consonent : " << consonent << " WhiteSpace : " << whiteSpace << endl;
}
// passing in function
pair<int, int> minMax(int arr[], int n)
{
    int mini = INT_MAX;
    int maxi = INT_MIN;
    for (int i = 0; i < n; i++)
    {
        int value = arr[i];
        maxi = max(value, maxi);
        mini = min(value, mini);
    }
    return {maxi, mini};
}
// Non returing fuction with no paratemer
void greet()
{
    cout << "Good Morning Mitron" << endl;
}

// Returning Function with paramters
int add(int a, int b)
{
    return (a + b);
}
int main()
{
    cout << "Practicing the basic concepts of c++";
    cout << endl;

    // printing or output
    cout << "Learning basics in c++ \n";

    // variable declaration
    int a;
    string b;
    float c;
    double d;
    char e;
    bool f;
    cout << a << b << c << d << e << f;

    /*User input */
    cout << "\nEnter values ";
    // cin >> a >> b >> c >> d >> e >> f;
    cout << "After user input:\n"
         << a << b << c << d << e << f;

    /*   Function declaration , types , intialization and passing parameters , function calling  */
    cout << endl;
    // function call
    greet();
    cout << add(10, 20);

    // IF -ELSE BLOCKS AND STATEMENT
    if (a >= 18)
    {
        cout << "\n You are an adult ";
    }
    else
    {
        cout << "\nYou are a Minor ";
    }

    /*LOOPS CONCEPT */

    // For loop

    cout << "Table of 2 :->";
    for (int i = 1; i <= 10; i++)
    {
        cout << (2 * i) << " ";
    }

    // while loop
    int i = 2;
    cout << "\n even number printing ";
    while (i <= 10)
    { // even number printing
        cout << i << " ";
        i += 2;
    }

    // do while loop
    do
    {
        // minimum one time execution always occur
        cout << "\nexecuted";

    } while (false);

    /* ARRAYS CONCEPT */

    // Array declaration
    int arr[5];
    int arr1[] = {2, 4, 5, 8}; // Arrays can be accessed via the indexed

    int n = sizeof(arr1) / sizeof(int);
    cout << "\nSize of array is : " << n << endl;

    pair<int, int> ans = minMax(arr1, n);
    cout << ans.first << "," << ans.second << endl;

    /* STRING CONCEPT */
    string str = "Yash Munde";
    cout << str << endl;
    // size finding
    int N = str.size();
    N = str.length(); // both are valid for finding the string length
    cout << N << endl;

    // string passing as arguments
    countVowelConsonent(str);
    cout << reverseString(str) << endl;

    /* CONCEPT OF RECURSION */
    cout << "Factorial of 5 is : " << factorial(5) << endl;

    // fibonacci series printing
    fiboPrint(5);

    return 0;
}
